reader_compare.go

```diff
@@ -0,0 +1,419 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package ionhash
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+
+	"github.com/amzn/ion-go/ion"
+)
+
+func Compare(reader1 ion.Reader, reader2 ion.Reader) error {
+	for true {
+		next, err := HasNext(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		if !next {
+			break
+		}
+
+		ionType1 := reader1.Type()
+		ionType2 := reader2.Type()
+
+		if ionType1 != ionType2 {
+			return fmt.Errorf("expected ion types to match")
+		}
+
+		ionHashReader, ok := reader1.(*hashReader)
+		if ok && ionHashReader.isInStruct() {
+			err := CompareFieldNames(reader1, reader2)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = CompareAnnotations(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		err = CompareAnnotationSymbols(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		err = CompareHasAnnotations(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		isNull1 := reader1.IsNull()
+		isNull2 := reader2.IsNull()
+		if isNull1 != isNull2 {
+			return fmt.Errorf("expected readers to have same value for IsNull()")
+		}
+
+		switch ionType1 {
+		case ion.NullType:
+			if !isNull1 {
+				return fmt.Errorf("expected ionType1 to be null")
+			}
+			if !isNull2 {
+				return fmt.Errorf("expected ionType2 to be null")
+			}
+			break
+		case ion.BoolType, ion.IntType, ion.FloatType, ion.DecimalType, ion.TimestampType,
+			ion.StringType, ion.SymbolType, ion.BlobType, ion.ClobType:
+			err := CompareScalars(ionType1, isNull1, reader1, reader2)
+			if err != nil {
+				return err
+			}
+			break
+		case ion.StructType, ion.ListType, ion.SexpType:
+			err := reader1.StepIn()
+			if err != nil {
+				return err
+			}
+
+			err = reader2.StepIn()
+			if err != nil {
+				return err
+			}
+
+			err = Compare(reader1, reader2)
+			if err != nil {
+				return err
+			}
+
+			err = reader1.StepOut()
+			if err != nil {
+				return err
+			}
+
+			err = reader2.StepOut()
+			if err != nil {
+				return err
+			}
+			break
+		default:
+			return &InvalidIonTypeError{ionType1}
+		}
+	}
+
+	next, err := HasNext(reader1, reader2)
+	if err != nil {
+		return err
+	}
+
+	if next {
+		return fmt.Errorf("expected HasNext() to return false")
+	}
+
+	return nil
+}
+
+func HasNext(reader1 ion.Reader, reader2 ion.Reader) (bool, error) {
+	more := reader1.Next()
+
+	if more != reader2.Next() {
+		return false, fmt.Errorf("next results don't match")
```

Comment:
Might want to add more context here, show the 2 boolean results for each reader.

Reply:
Done.